### PR TITLE
internal: add TurboShake{128,256}

### PR DESCRIPTION
**Pull request by bwesterb**
```
goos: darwin
goarch: arm64
pkg: github.com/cloudflare/circl/internal/sha3
BenchmarkPermutationFunctionTurbo-12    	 7648284	       145.6 ns/op	1374.07 MB/s
BenchmarkPermutationFunction-12         	 4152145	       289.1 ns/op	 691.77 MB/s
BenchmarkSha3_512_MTU-12                	  200443	      5956 ns/op	 226.66 MB/s
BenchmarkSha3_384_MTU-12                	  277851	      4302 ns/op	 313.79 MB/s
BenchmarkSha3_256_MTU-12                	  347239	      3416 ns/op	 395.19 MB/s
BenchmarkSha3_224_MTU-12                	  370054	      3258 ns/op	 414.40 MB/s
BenchmarkShake128_MTU-12                	  420195	      2842 ns/op	 475.10 MB/s
BenchmarkShake256_MTU-12                	  390090	      3069 ns/op	 439.88 MB/s
BenchmarkShake256_16x-12                	   29004	     41288 ns/op	 396.82 MB/s
BenchmarkShake256_1MiB-12               	     506	   2360097 ns/op	 444.29 MB/s
BenchmarkTurboShake128_1MiB-12          	    1160	   1026859 ns/op	1021.15 MB/s
BenchmarkTurboShake256_1MiB-12          	     951	   1253763 ns/op	 836.34 MB/s
BenchmarkSha3_512_1MiB-12               	     274	   4353505 ns/op	 240.86 MB/s
PASS
ok  	github.com/cloudflare/circl/internal/sha3	17.678s
```

### CODE REVIEW

File: simd/keccakf1600/f1600x.go [hunk 1]
```diff
@@ -104,46 +102,54 @@ func (s *StateX2) Initialize() []uint64 {
...
 		for j := 0; j < 25; j++ {
 			a[2*j+i] = buf[j]
 		}
 	}
 }
+
+var enabledX2 bool
+
+func init() {
+	enabledX2 = runtime.GOARCH == "arm64" && runtime.GOOS == "darwin"
```
Comment (armfazh) [inline simd/keccakf1600/f1600x.go:154]: does the code is OS-dependent? we only need to check whether the NEON/ASIMD instructions are supported right?

**Review by armfazh — CHANGES_REQUESTED** [review]

File: simd/keccakf1600/f1600x.go (hunk 1 above)
Comment (bwesterb) [reply simd/keccakf1600/f1600x.go:154]: No, but any Mac on ARM64 supports these extensions. We're just waiting for Go to add the feature flag, which they're refusing to do for the moment: https://github.com/golang/go/pull/46300

File: simd/keccakf1600/f1600x.go [hunk 2]
```diff
@@ -104,46 +102,54 @@ func (s *StateX2) Initialize() []uint64 {
 
 // Permute performs the four parallel Keccak-f[1600]s interleaved on the slice
 // returned from Initialize().
-func (s *StateX4) Permute() {
+// If turbo is true, applies 12-round variant instead of the usual 24.
+func (s *StateX4) Permute(turbo bool) {
```
[the remark this answers is not in the document]
Comment (bwesterb) [reply simd/keccakf1600/f1600x.go:106]: Will do. First I'm finishing K12 :)

[the remark this answers is not in the document]
Comment (bwesterb) [reply simd/keccakf1600/f1600x.go]: Done

**Comment by bwesterb** [comment]
@armfazh I addressed the comments. Please have another look.

File: internal/sha3/shake.go [hunk 3]
```diff
@@ -57,13 +57,35 @@ func NewShake128() State {
...
 // Its generic security strength is 256 bits against all attacks if
 // at least 64 bytes of its output are used.
 func NewShake256() State {
 	return State{rate: rate256, dsbyte: dsbyteShake}
 }
 
+// NewTurboShake256 creates a new TurboSHAKE256 variable-output-length ShakeHash.
+// Its generic security strength is 256 bits against all attacks if
+// at least 64 bytes of its output are used.
+// D is the seperationByte and must be between 0x01 and 0x7f inclusive.
```
Comment (armfazh) [inline internal/sha3/shake.go:81]:
```suggestion
// D is the domain separation parameter and must be between 0x01 and 0x7f inclusive.
```

File: internal/sha3/shake.go [hunk 4]
```diff
@@ -57,13 +57,35 @@ func NewShake128() State {
 	return State{rate: rate128, dsbyte: dsbyteShake}
 }
 
+// NewTurboShake128 creates a new TurboSHAKE128 variable-output-length ShakeHash.
+// Its generic security strength is 128 bits against all attacks if at
+// least 32 bytes of its output are used.
+// D is the seperationByte and must be between 0x01 and 0x7f inclusive.
```
Comment (armfazh) [inline internal/sha3/shake.go:63]:
```suggestion
// D is the domain separation parameter and must be between 0x01 and 0x7f inclusive.
```

File: simd/keccakf1600/f1600x2_arm64.s [hunk 5]
```diff
@@ -19,6 +19,12 @@ TEXT ·f1600x2ARM(SB), NOSPLIT, $0-16
     VLD1.P 64(R0), [V20.B16, V21.B16, V22.B16, V23.B16]
     VLD1.P (R0),   [V24.B16]
 
+    MOVBU turbo+16(FP), R4
+    CBZ R4, loop
+
+    SUBS $12, R3, R3
```
Comment (armfazh) [inline simd/keccakf1600/f1600x2_arm64.s:25]: Does the `S` is really needed?

File: simd/keccakf1600/internal/asm/go.sum [hunk 6]
```diff
@@ -1,24 +1,68 @@
 github.com/mmcloughlin/avo v0.0.0-20200523190732-4439b6b2c061 h1:UCU8+cLbbvyxi0sQ9fSeoEhZgvrrD9HKMtX6Gmc1vk8=
```
Comment (armfazh) [inline simd/keccakf1600/internal/asm/go.sum:1]: try doing a `go mod tidy` to update this file.

**Review by armfazh — APPROVED** [review]
lgtm, just fix the nits.

**Comment by bwesterb** [comment]
> lgtm, just fix the nits.

Thanks. I addressed the comments.

**Review by armfazh — APPROVED** [review]

**Pull request merged**